[clang] Implement pointer authentication for C++ virtual functions, v-tables, and VTTs

ojhunt opened the pull request:
Virtual function pointer entries in v-tables are signed with address discrimination in addition to declaration-based discrimination, where an integer discriminator the string hash (see `ptrauth_string_discriminator`) of the mangled name of the overridden method.  This notably provides diversity based on the full signature of the overridden method, including the method name and parameter types.
This patch introduces ItaniumVTableContext logic to find the original declaration of the overridden method.  On AArch64, these pointers are signed using the `IA` key (the process-independent code key.)

V-table pointers can be signed with either no discrimination, or a similar scheme using address and decl-based discrimination.  In this case, the integer discriminator is the string hash of the mangled v-table identifier of the class that originally introduced the vtable pointer.  On AArch64, these pointers are signed using the `DA` key (the process-independent data key.)

Not using discrimination allows attackers to simply copy valid v-table pointers from one object to another.  However, using a uniform discriminator of 0 does have positive performance and code-size implications on AArch64, and diversity for the most important v-table access pattern (virtual dispatch) is already better assured by the signing schemas used on the virtual functions.  It is also known that some code in practice copies objects containing v-tables with `memcpy`, and while this is not permitted formally, it is something that may be invasive to eliminate.

This is controlled by:
```
  -fptrauth-vtable-pointer-type-discrimination
  -fptrauth-vtable-pointer-address-discrimination
```

In addition, this provides fine-grained controls in the ptrauth_vtable_pointer attribute, which allows overriding the default ptrauth schema for vtable pointers on a given class hierarchy, e.g.:
```
  [[clang::ptrauth_vtable_pointer(no_authentication, no_address_discrimination, 
                                  no_extra_discrimination)]]
  [[clang::ptrauth_vtable_pointer(default_key, default_address_discrimination,
                                  custom_discrimination, 0xf00d)]]
```
The override is then mangled as a parametrized vendor extension:
```
"__vtptrauth" I
 <key>
 <addressDiscriminated>
 <extraDiscriminator>
E
```
To support this attribute, this patch adds a small extension to the attribute-emitter tablegen backend.

Note that there are known areas where signing is either missing altogether or can be strengthened.  Some will be addressed in later changes (e.g., member function pointers, some RTTI).
`dynamic_cast` in particular is handled by emitting an artificial v-table pointer load (in a way that always authenticates it) before the runtime call itself, as the runtime doesn't have enough information today to properly authenticate it.  Instead, the runtime is currently expected to strip the v-table pointer.

ojhunt committed:
[clang] Record the original declaration for which a v-table slot was created.

Co-Authored-By: John McCall <rjmccall@apple.com>

ojhunt committed:
[clang] Implement pointer authentication for C++ virtual functions, v-tables, and VTTs.

This also implements the ptrauth_vtable_pointer attribute to allow
overriding the default ptrauth schema for vtable pointers.

Co-Authored-By: John McCall <rjmccall@apple.com>

Endilll reviewed (commented):
`Sema.h` changes look good.

asl commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -1261,6 +1262,10 @@ class ASTContext : public RefCountedBase<ASTContext> {
   /// space.
   QualType removeAddrSpaceQualType(QualType T) const;
 
+  /// Return the "other" type-specific discriminator for the given type.
```
The comment here looks a bit misleading without knowing what function is doing. Could it be reformulated in some better way?

asl replied:
@ojhunt This looks much better, thanks!

ojhunt committed:
Respond to review comments

ojhunt commented:
Had to do a force push to resolve merge conflicts following @ahatanak's PR, so this change now includes `CodeGenFunction::EmitPointerAuthAuth` etc

asl commented on clang/include/clang/Basic/Thunk.h:
```diff
@@ -162,20 +162,27 @@ struct ThunkInfo {
   ThunkInfo() : Method(nullptr) {}
 
   ThunkInfo(const ThisAdjustment &This, const ReturnAdjustment &Return,
+            const Type *thisType,
```
nit: consistent naming of parameters

asl commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -3088,6 +3089,17 @@ QualType ASTContext::removeAddrSpaceQualType(QualType T) const {
+  SmallString<256> Str;
+  llvm::raw_svector_ostream Out(Str);
+  MC->mangleCXXVTable(record, Out);
+  return llvm::getPointerAuthStableSipHash16(Str.c_str());
```
```suggestion
  return llvm::getPointerAuthStableSipHash16(Str.str());
```

asl commented on clang/lib/AST/ItaniumMangle.cpp:
```diff
@@ -7037,8 +7036,64 @@ void ItaniumMangleContextImpl::mangleCXXDtorComdat(const CXXDestructorDecl *D,
+  manglerStream << "I";
+  if (auto explicitAuth =
+          ptrauthClassRecord->getAttr<VTablePointerAuthenticationAttr>()) {
+    manglerStream << "Lj" << explicitAuth->getKey();
```
Do we have mangling bits documented somewhere?

ojhunt replied:
I'll see what I can find

asl commented on clang/lib/AST/ItaniumMangle.cpp:
```diff
@@ -7064,21 +7119,29 @@ void ItaniumMangleContextImpl::mangleThunk(const CXXMethodDecl *MD,
                            ThisAdjustment.Virtual.Itanium.VCallOffsetOffset);
 
   Mangler.mangleFunctionEncoding(GlobalDecl(DD, Type));
+  if (!elideOverrideInfo)
```
update comment above with mangling scheme description?

asl commented on clang/lib/AST/VTableBuilder.cpp:
```diff
@@ -3094,9 +3188,10 @@ void VFTableBuilder::AddMethods(BaseSubobject Base, unsigned BaseDepth,
       }
     }
-
+    auto thisType = (OverriddenMD ? OverriddenMD : MD)->getThisType().getTypePtr();
```
```suggestion
    auto ThisType = (OverriddenMD ? OverriddenMD : MD)->getThisType().getTypePtr();
```

asl commented on clang/lib/AST/VTableBuilder.cpp:
```diff
@@ -3094,9 +3188,10 @@ void VFTableBuilder::AddMethods(BaseSubobject Base, unsigned BaseDepth,
+    auto thisType = (OverriddenMD ? OverriddenMD : MD)->getThisType().getTypePtr();
     AddMethod(FinalOverriderMD,
               ThunkInfo(ThisAdjustmentOffset, ReturnAdjustment,
+                        thisType,
```
```suggestion
                        ThisType,
```

asl commented on clang/lib/CodeGen/CGClass.cpp:
```diff
@@ -2588,6 +2588,12 @@ void CodeGenFunction::InitializeVTablePointer(const VPtr &Vptr) {
   // the same addr space. Note that this might not be LLVM address space 0.
   VTableField = VTableField.withElementType(PtrTy);
 
+  if (auto authenticationInfo = CGM.getVTablePointerAuthInfo(
```
```suggestion
  if (auto AuthenticationInfo = CGM.getVTablePointerAuthInfo(
```

asl commented on clang/lib/CodeGen/CGClass.cpp:
```diff
@@ -2588,6 +2588,12 @@ void CodeGenFunction::InitializeVTablePointer(const VPtr &Vptr) {
+  if (auto authenticationInfo = CGM.getVTablePointerAuthInfo(
+          this, Vptr.Base.getBase(), VTableField.emitRawPointer(*this))) {
+    VTableAddressPoint =
+        EmitPointerAuthSign(*authenticationInfo, VTableAddressPoint);
```
```suggestion
        EmitPointerAuthSign(*AuthenticationInfo, VTableAddressPoint);
```

asl commented on clang/lib/CodeGen/CGClass.cpp:
```diff
@@ -2681,12 +2687,35 @@ void CodeGenFunction::InitializeVTablePointers(const CXXRecordDecl *RD) {
   TBAAAccessInfo TBAAInfo = CGM.getTBAAVTablePtrAccessInfo(VTableTy);
   CGM.DecorateInstructionWithTBAA(VTable, TBAAInfo);
 
+  if (auto authenticationInfo =
```
```suggestion
  if (auto *AuthenticationInfo =
```

asl commented on clang/lib/CodeGen/CGClass.cpp:
```diff
@@ -2681,12 +2687,35 @@ void CodeGenFunction::InitializeVTablePointers(const CXXRecordDecl *RD) {
                                            llvm::Type *VTableTy,
-                                           const CXXRecordDecl *RD) {
+                                           const CXXRecordDecl *RD,
+                                           VTableAuthMode authMode) {
```
```suggestion
                                           VTableAuthMode AuthMode) {
```

asl commented on clang/lib/CodeGen/CGClass.cpp:
```diff
@@ -2681,12 +2687,35 @@ void CodeGenFunction::InitializeVTablePointers(const CXXRecordDecl *RD) {
 
+  if (auto authenticationInfo =
+            CGM.getVTablePointerAuthInfo(this, RD, This.emitRawPointer(*this))) {
+    if (authMode != VTableAuthMode::UnsafeUbsanStrip) {
```
```suggestion
    if (AuthMode != VTableAuthMode::UnsafeUbsanStrip) {
```

asl commented on clang/lib/CodeGen/CGClass.cpp:
```diff
@@ -2681,12 +2687,35 @@ void CodeGenFunction::InitializeVTablePointers(const CXXRecordDecl *RD) {
+            CGM.getVTablePointerAuthInfo(this, RD, This.emitRawPointer(*this))) {
+    if (authMode != VTableAuthMode::UnsafeUbsanStrip) {
+      VTable = cast<llvm::Instruction>(
+          EmitPointerAuthAuth(*authenticationInfo, VTable));
```
```suggestion
          EmitPointerAuthAuth(*AuthenticationInfo, VTable));
```

asl commented on clang/lib/CodeGen/CGClass.cpp:
```diff
@@ -2681,12 +2687,35 @@ void CodeGenFunction::InitializeVTablePointers(const CXXRecordDecl *RD) {
+    if (authMode != VTableAuthMode::UnsafeUbsanStrip) {
+      VTable = cast<llvm::Instruction>(
+          EmitPointerAuthAuth(*authenticationInfo, VTable));
+      if (authMode == VTableAuthMode::MustTrap) {
```
```suggestion
      if (AuthMode == VTableAuthMode::MustTrap) {
```

asl commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -803,6 +803,14 @@ bool ConstStructBuilder::Build(const APValue &Val, const RecordDecl *RD,
       llvm::Constant *VTableAddressPoint =
           CGM.getCXXABI().getVTableAddressPoint(BaseSubobject(CD, Offset),
                                                 VTableClass);
+      if (auto authentication =
```
```suggestion
      if (auto *Authentication =
```

asl commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -803,6 +803,14 @@ bool ConstStructBuilder::Build(const APValue &Val, const RecordDecl *RD,
+      if (auto authentication =
+              CGM.getVTablePointerAuthentication(VTableClass)) {
+        VTableAddressPoint = Emitter.tryEmitConstantSignedPointer(
+            VTableAddressPoint, *authentication);
```
```suggestion
            VTableAddressPoint, *Authentication);
```

asl commented on clang/lib/CodeGen/CGPointerAuth.cpp:
```diff
@@ -23,11 +23,53 @@
 
+/// Given a pointer-authentication schema, return a concrete "other"
+/// discriminator for it.
+llvm::Constant *CodeGenModule::getPointerAuthOtherDiscriminator(
```
It would be great to have consistent code style for variable names in this file.

asl commented on clang/lib/CodeGen/CGVTT.cpp:
```diff
@@ -90,6 +90,11 @@ CodeGenVTables::EmitVTTDefinition(llvm::GlobalVariable *VTT,
      llvm::Constant *Init = llvm::ConstantExpr::getGetElementPtr(
          VTable->getValueType(), VTable, Idxs, /*InBounds=*/true, InRange);
 
+     if (auto &schema =
```
```suggestion
     if (auto &Schema =
```

asl commented on clang/lib/CodeGen/CGVTT.cpp:
```diff
@@ -90,6 +90,11 @@ CodeGenVTables::EmitVTTDefinition(llvm::GlobalVariable *VTT,
 
+     if (auto &schema =
+             CGM.getCodeGenOpts().PointerAuth.CXXVTTVTablePointers)
+       Init = CGM.getConstantSignedPointer(Init, schema, nullptr, GlobalDecl(),
```
```suggestion
       Init = CGM.getConstantSignedPointer(Init, Schema, nullptr, GlobalDecl(),
```

asl commented on clang/lib/CodeGen/CGVTables.cpp:
```diff
@@ -219,8 +219,10 @@ CodeGenFunction::GenerateVarArgsThunk(llvm::Function *Fn,
-  llvm::Value *AdjustedThisPtr =
-      CGM.getCXXABI().performThisAdjustment(*this, ThisPtr, Thunk.This);
+
+  const CXXRecordDecl *thisValueClass = Thunk.ThisType->getPointeeCXXRecordDecl();
```
```suggestion
  const CXXRecordDecl *ThisValueClass = Thunk.ThisType->getPointeeCXXRecordDecl();
```

asl commented on clang/lib/CodeGen/CGVTables.cpp:
```diff
@@ -219,8 +219,10 @@ CodeGenFunction::GenerateVarArgsThunk(llvm::Function *Fn,
+
+  const CXXRecordDecl *thisValueClass = Thunk.ThisType->getPointeeCXXRecordDecl();
+  llvm::Value *AdjustedThisPtr = CGM.getCXXABI().performThisAdjustment(
+      *this, ThisPtr, thisValueClass, Thunk);
```
```suggestion
      *this, ThisPtr, ThisValueClass, Thunk);
```

asl commented on clang/lib/CodeGen/CGVTables.cpp:
```diff
@@ -307,10 +309,15 @@ void CodeGenFunction::EmitCallAndReturnForThunk(llvm::FunctionCallee Callee,
   const CXXMethodDecl *MD = cast<CXXMethodDecl>(CurGD.getDecl());
 
   // Adjust the 'this' pointer if necessary
+  const CXXRecordDecl *thisValueClass =
```
```suggestion
  const CXXRecordDecl *ThisValueClass =
```

asl commented on clang/lib/CodeGen/CGVTables.cpp:
```diff
@@ -307,10 +309,15 @@ void CodeGenFunction::EmitCallAndReturnForThunk(llvm::FunctionCallee Callee,
+  const CXXRecordDecl *thisValueClass =
+      MD->getThisType()->getPointeeCXXRecordDecl();
+  if (Thunk)
+    thisValueClass = Thunk->ThisType->getPointeeCXXRecordDecl();
```
```suggestion
    ThisValueClass = Thunk->ThisType->getPointeeCXXRecordDecl();
```

asl commented on clang/lib/CodeGen/CGVTables.cpp:
```diff
@@ -307,10 +309,15 @@ void CodeGenFunction::EmitCallAndReturnForThunk(llvm::FunctionCallee Callee,
-                          *this, LoadCXXThisAddress(), Thunk->This)
-          : LoadCXXThis();
+      Thunk ? CGM.getCXXABI().performThisAdjustment(*this, LoadCXXThisAddress(),
+                                                    thisValueClass, *Thunk)
```
```suggestion
                                                    ThisValueClass, *Thunk)
```

asl commented on clang/lib/CodeGen/CGVTables.cpp:
```diff
@@ -841,6 +866,9 @@ void CodeGenVTables::addVTableComponent(ConstantArrayBuilder &builder,
       if (FnAS != GVAS)
         fnPtr =
             llvm::ConstantExpr::getAddrSpaceCast(fnPtr, CGM.GlobalsInt8PtrTy);
+      if (auto &schema =
```
```suggestion
      if (const auto &Schema =
```

asl commented on clang/lib/CodeGen/CGVTables.cpp:
```diff
@@ -841,6 +866,9 @@ void CodeGenVTables::addVTableComponent(ConstantArrayBuilder &builder,
             llvm::ConstantExpr::getAddrSpaceCast(fnPtr, CGM.GlobalsInt8PtrTy);
+      if (auto &schema =
+          CGM.getCodeGenOpts().PointerAuth.CXXVirtualFunctionPointers)
+        return builder.addSignedPointer(fnPtr, schema, GD, QualType());
```
```suggestion
        return builder.addSignedPointer(fnPtr, Schema, GD, QualType());
```

asl commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -3063,3 +3063,66 @@ void CodeGenFunction::EmitPointerAuthOperandBundle(
   bundles.emplace_back("ptrauth", args);
 }
+
+static llvm::Value *EmitPointerAuthCommon(CodeGenFunction &CGF,
```
ditto about code style

asl commented on clang/lib/CodeGen/CodeGenFunction.h:
```diff
@@ -2456,10 +2456,20 @@ class CodeGenFunction : public CodeGenTypeCache {
+  llvm::Value *
+  GetVTablePtr(Address This, llvm::Type *VTableTy,
+               const CXXRecordDecl *VTableClass,
+               VTableAuthMode authMode = VTableAuthMode::Authenticate);
```
```suggestion
               VTableAuthMode AuthMode = VTableAuthMode::Authenticate);
```

asl commented on clang/lib/CodeGen/CodeGenFunction.h:
```diff
@@ -4406,10 +4416,19 @@ class CodeGenFunction : public CodeGenTypeCache {
   bool isPointerKnownNonNull(const Expr *E);
+
+  /// Create the discriminator from the storage address and the entity hash.
+  llvm::Value *EmitPointerAuthBlendDiscriminator(llvm::Value *storageAddress,
```
```suggestion
  llvm::Value *EmitPointerAuthBlendDiscriminator(llvm::Value *StorageAddress,
```

asl commented on clang/lib/CodeGen/CodeGenFunction.h:
```diff
@@ -4406,10 +4416,19 @@ class CodeGenFunction : public CodeGenTypeCache {
+
+  /// Create the discriminator from the storage address and the entity hash.
+  llvm::Value *EmitPointerAuthBlendDiscriminator(llvm::Value *storageAddress,
+                                                 llvm::Value *discriminator);
```
```suggestion
                                                 llvm::Value *Discriminator);
```

Endilll commented:
@asl It would be nice if you submit the next round of review as a single review, instead of 29 individual comments.

asl commented:
> @asl It would be nice if you submit the next round of review as a single review, instead of 29 individual comments.

Oh, strange. I thought it was a single review...

ojhunt committed:
Update clang/lib/AST/VTableBuilder.cpp

Co-authored-by: Anton Korobeynikov <anton@korobeynikov.info>

ojhunt committed:
Address review comments

asl reviewed (approved):
Thanks!

ojhunt commented on clang/lib/CodeGen/ConstantInitBuilder.cpp:
```diff
@@ -296,3 +296,21 @@ ConstantAggregateBuilderBase::finishStruct(llvm::StructType *ty) {
   buffer.erase(buffer.begin() + Begin, buffer.end());
   return constant;
 }
+
```
@asl Updating to resolve conflict required bringing this function in that was previously in Akira's work

asl replied:
Yeah, LGTM!

ojhunt committed:
Update code style

ahmedbougacha committed:
Update following rebase.

- update ubsan vptr codegen/tests for hash/mixer change
- adopt renamed siphash function; pass SmallString as StringRef
- bring back removed getConstantSignedPointer declaration
- move shouldSignPointer into CGM
- remove unneeded addSignedPointer variant
- move other addSignedPointer variant back to ConstantInitBuilder

ahmedbougacha committed:
Drop no-longer-necessary cast-stripping helper.

If it wasn't for inttoptr/ptrtoint, this probably could have been
stripPointerCasts anyway.

Given the sole existing user now only produces plain ConstantPtrAuth
values anyway (unlike the previous dance with llvm.ptrauth global),
I don't think this is needed anymore.

ahmedbougacha committed:
Do a final pass to address cosmetics.

ahmedbougacha committed:
Make test ptrauth-explicit-vtable-pointer-control.cpp readable.

Or at least, a little bit more readable.

AaronBallman commented:
It was observed in https://github.com/llvm/llvm-project/pull/99993/ that this PR introduced tablegen code that appears to be unnecessary (`attributeHasStrictIdentifierArgAtIndex()`), and that makes me worry that these changes are not correct (certainly they lack test coverage).

What was the intent behind that code? Should it be removed or did we discover a bug with this PR?

CC @ojhunt @ahmedbougacha @asl
